### ISSUE TITLE
A new activation function ACON that is very simple and effective !!

### ISSUE THREAD

**Issue by nmaac** [ultralytics/yolov5#2891]
## 🚀 Feature

There is a new activation function [ACON (CVPR 2021)](https://arxiv.org/pdf/2009.04759.pdf) that unifies ReLU and Swish. 
ACON is simple but very effective, code is here: https://github.com/nmaac/acon/blob/main/acon.py#L19

![image](https://user-images.githubusercontent.com/5032208/115676962-a38dfe80-a382-11eb-9883-61fa3216e3e6.png)

The improvements are very significant:
![image](https://user-images.githubusercontent.com/5032208/115680180-eac9be80-a385-11eb-9c7a-8643db552c69.png)

## Pitch

I would like to suggest replacing SiLU with ACON directly because SiLU (Swish) is used in your project, its general and effective form ACON may also show improvements.

## Alternatives

It also has an enhanced version meta-ACON that uses a small network to learn beta explicitly, which may influence the speed a bit.

## Additional context

[Code](https://github.com/nmaac/acon) and [paper](https://arxiv.org/pdf/2009.04759.pdf).

**Comment by ilem777** [comment]
looking forward to adding in!

**Comment by glenn-jocher** [comment]
@nmaac thanks for the idea, looks promising! Any object detection results so far?

**Comment by nmaac** [comment]
There are some detection results:

![image](https://user-images.githubusercontent.com/5032208/115687713-08e6ed00-a38d-11eb-8065-302ec1e6d451.png)

I did not test it on yolov5, but seems it has the potential to make nearly cost-free improvements, by simply replacing SiLU.

**Comment by glenn-jocher** [comment]
@nmaac ah great, thank you! Yes this is quite a significant improvement in your Table 9. Which ACON version would you recommend we try, and what values for p1, p2, Beta? 

- meta-ACON
- ACON-A
- ACON-B
- ACON-C

The right place to include a new activation would be utils/activations, and then the place to swap out nn.SiLU() for a new activation is here on L39 of models/common.py
https://github.com/ultralytics/yolov5/blob/d48a34dca72c7df1f684a1ff33c18beebc0f0ed9/models/common.py#L33-L43

**Comment by nmaac** [comment]
I would like to suggest ACON-C, which improves accuracy without a negligible overhead.

You can use the code directly:  

https://github.com/nmaac/acon/blob/8782b65f5d7b3523f656beceb586b54d04019705/acon.py#L4-19

**Comment by ilem777** [comment]
What are the results of your tests? In terms of performance and inference time, how does the ACON activation function compare with the built-in activation function of YOLOv5? @nmaac  @glenn-jocher

**Comment by glenn-jocher** [comment]
@nmaac @ilem777 I've added AconC to our activations study here:
https://wandb.ai/glenn-jocher/activations

I just started runs with **AconC()**, **MetaAconC()** and **FReLU()**, you can track their progress live at the link above. Training time will be about 3 days. I tried MetaAconC but ran into issues. The nn.batchnorm2d(16) layers produced errors on inputs of size (1,16,1,1), perhaps I implemented the function incorrectly.

**Comment by glenn-jocher** [comment]
@AyushExel I spotted something concerning I was hoping you could look at. When runs are public, like the activation study above, the 'stop run' button appears to work even when the visitor is incognito / no signin.

**Comment by AyushExel** [comment]
@glenn-jocher thanks for reporting this. I'll check if the button for non-authorized users actually stops the runs or not. If it does then it's a very bad bug otherwise it's just a minor frontend bug. I'll file a ticket for this to get fixed

**Comment by WongKinYiu** [comment]
it because nn.batchnorm2d need batch size > 1 when training.
the simplest way to solve the problem is change [this line](https://github.com/ultralytics/yolov5/blob/master/models/yolo.py#L93) to
```python
            m.stride = torch.tensor([s / x.shape[-2] for x in self.forward(torch.zeros(2, ch, s, s))])  # forward
```

**Comment by nmaac** [comment]
@glenn-jocher you can simply remove the two bn
layers in MetaAcon which does not affect the accuracy much.

**Comment by glenn-jocher** [comment]
@nmaac oh, I think I misunderstood before. I think you mean to remove self.bn1 and self.bn2 completely from the MetaAconC() module for all batch-sizes?

**Comment by glenn-jocher** [comment]
@WongKinYiu yes, this is a good solution too, though will make model creation a bit slower for all other models. The nn.batchnorm2d() layers are ok for batch-size 1 inference?

**Comment by glenn-jocher** [comment]
@WongKinYiu @nmaac I'm curious, looking at the ACON implementation have you guys tried simply training with SiLU with Beta? I've never done this before. nn.SiLU() does not allow this but I think I might try testing this using a custom SiLU to see how this affects the results.

**Comment by WongKinYiu** [comment]
nn.batchnorm2d() layers can do batch-size 1 inference.
or instancenorm is another choice.

**Comment by nmaac** [comment]
@glenn-jocher  SiLU with beta does not show benefits, in the [paper](https://arxiv.org/abs/1710.05941) Swish-1 and Swish show comparable results when set beta=1, specifically,
Swish: x\*sigmoid(beta\*x)
Swish-1(SiLU): x\*sigmoid(x)

Therefore meta-ACON uses an explicitly way to learn beta which show the improvements.

**Comment by glenn-jocher** [comment]
@nmaac understood, thanks for explanation. I had to completely remove the BN layers from MetaAconC otherwise instabilities appeared in the training (two 'STOPPED' runs below). Results should be done in about a day, but based on the current trends it doesn't initially seem like I was able to produce better results with either AconC or MetaAconC. The best performing activation in the study by far was FReLU, though this should be taken with a grain of salt as FReLU is really blurring the lines between an activation and a convolution layer. Due to the added parameters and FLOPS I would also assume FReLU would disproportionately improve smaller models like YOLOv5s, with unclear correlation to improving larger models like YOLOv5x6, which may necessitate a second study in the future.
https://wandb.ai/glenn-jocher/activations

**Comment by nmaac** [comment]
@glenn-jocher Yes the curves show comparable results. Which activations did you change? Did you pre-train the backbone? In my experiments I usually change the activations in the backbone and pre-train the backbone on ImageNet first. I can help with the pre-training if needed :)

**Comment by WongKinYiu** [comment]
i think the main reason is that @glenn-jocher forget to add `p1`, `p2`, and `(beta)` into no decay optimized group.
https://github.com/nmaac/acon/blob/8782b65f5d7b3523f656beceb586b54d04019705/ACON/ResNet_ACON/utils.py#L82

**Comment by glenn-jocher** [comment]
@nmaac well that's a good question, should the activation function parameters be exempt from weight decay? We use the following parameter groups to exempt .bias parameters and BatchNorm layers from weight decay, so at the moment only the fc1, fc2 biases are exempt from decay. 
https://github.com/ultralytics/yolov5/blob/184991672636838453e796f72268833dff788d07/train.py#L115-L123

The activation function implementations are all in utils/activations:
https://github.com/ultralytics/yolov5/blob/184991672636838453e796f72268833dff788d07/utils/activations.py#L58-L98

The **activations_study branch** used in this study replaces **all activations** in the YOLOv5 model by redefining `self.act` here:
https://github.com/ultralytics/yolov5/blob/c9c95fb282322bc2928f44488f2467c5f4104f09/models/common.py#L34-L55

All YOLOv5 models are trained from scratch to 300 epochs using all default settings. Training commands are shown in the W&B link to reproduce (COCO dataset autodownloads):
```
train.py --batch 64 --data coco.yaml --cfg yolov5s.yaml --weights '' --epochs 300 --img 640 --project activations --name yolov5s-MetaAconC_noBN --device 0
```

**Comment by developer0hye** [comment]
@glenn-jocher 
Is there any progress on this issue?

**Comment by glenn-jocher** [comment]
@developer0hye well I'm not sure. The ACON authors @nmaac didn't answer my question of whether we should exempt some of the ACON parameters from weight decay. The current results are here for all the activations on YOLOv5s: https://wandb.ai/glenn-jocher/activations

**Comment by nmaac** [comment]
@glenn-jocher @developer0hye In my experiments, the weight decay setting does not affect the results very much.

But I suggest try another initialization approach:

> self.p1 = nn.Parameter(torch.normal(1, 0.01, size=(1, width, 1, 1)))
    self.p2 = nn.Parameter(torch.normal(0, 0.01, size=(1, width, 1, 1)))
    self.beta = nn.Parameter(torch.normal(1, 0.01, size=(1, width, 1, 1)))

**Comment by WongKinYiu** [comment]
@nmaac 

in my experiments:

1. original silu ~300 epochs: 51.9% AP

2. old init ~300 epochs: 50.6% AP 

3. new initial ~300 epochs: 51.5% AP
> self.p1 = nn.Parameter(torch.normal(1, 0.01, size=(1, width, 1, 1)))
> self.p2 = nn.Parameter(torch.normal(0, 0.01, size=(1, width, 1, 1)))
> self.beta = nn.Parameter(torch.normal(1, 0.01, size=(1, width, 1, 1)))

and old initial with decay drops 0.2% AP

**Comment by glenn-jocher** [comment]
@nmaac @WongKinYiu got it, thanks guys!

**Comment by iumyx2612** [comment]
Any updates on this? How's ACON
